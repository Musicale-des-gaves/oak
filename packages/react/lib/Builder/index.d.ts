import { ReactNode, ComponentPropsWithRef, FormEvent } from 'react';
import { AddonObject, BuilderOptions, ElementObject } from '@oakjs/core';

interface ImageUploadCallbackResult {
  url: string;
  name: string;
  [key: string]: any;
}

declare interface BuilderProps extends ComponentPropsWithRef<any> {
  className?: string;
  defaultValue?: Array<ElementObject>;
  value?: Array<ElementObject>;
  addons: Array<AddonObject>;
  rootBoundary?: string | Element | DocumentFragment;
  onChange?(content: Array<ElementObject>): void;
  onImageUpload?(event: FormEvent): Promise<ImageUploadCallbackResult>;
  children?: ReactNode | JSX.Element;
  activeTextSheet?: string;
  options?: BuilderOptions;
  [key: string]: any;
}

declare function Builder(props: BuilderProps): ReactNode | JSX.Element;

export default Builder;
