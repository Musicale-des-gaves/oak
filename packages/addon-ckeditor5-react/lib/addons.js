import ClassicEditor from '@oakjs/ckeditor5-build-custom';

import Field from './Field';

export const ckeditorField = ({ config, editor, ...props } = {}) => ({
  type: 'ckeditor',
  render: Field,
  ...props,
  props: {
    editor: editor || ClassicEditor,
    config: {
      ...config,
      link: {
        decorators: {
          openInNewTab: {
            mode: 'manual',
            label: 'Open in a new tab',
            defaultValue: false,
            attributes: {
              target: '_blank',
              rel: 'noopener noreferrer',
            },
          },
        },
      },
      toolbar: {
        items: [
          'undo',
          'redo',
          '|',
          'bold',
          'italic',
          'strikethrough',
          'underline',
          'link',
          'fontColor',
          'fontSize',
          'bulletedList',
          'numberedList',
          '|',
          'removeFormat',
          '|',
          'outdent',
          'indent',
          'alignment',
          '|',
          'horizontalLine',
          'blockQuote',
          'insertTable',
        ],
      },
    },
  },
});

export const ckeditorFieldAddon = props => ({
  fields: [ckeditorField(props)],
});
