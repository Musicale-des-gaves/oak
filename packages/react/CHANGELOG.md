# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [3.0.0](https://github.com/p3ol/oak/compare/v3.0.0-alpha.40...v3.0.0) (2023-05-15)


### Bug Fixes

* **react:** allow to paste multiple elements inside containers ([7efe183](https://github.com/p3ol/oak/commit/7efe183ad1cf57e8e1922110020591629294b727))
* update floating ui ([7c1f08f](https://github.com/p3ol/oak/commit/7c1f08f91c14a6377c8d0a4d8fb9668f6a5fe57e))
* update junipero to major ([8515b9f](https://github.com/p3ol/oak/commit/8515b9f397ab1572fc38ae62c607d87458e68c32))
* upgrade various dependencies ([b70b8bf](https://github.com/p3ol/oak/commit/b70b8bfb968147fe8021c2075d9e5f1a9b54242e))
