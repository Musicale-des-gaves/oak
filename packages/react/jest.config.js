const path = require('path');

module.exports = {
  displayName: '@oakjs/react',
  testEnvironment: 'jsdom',
  clearMocks: true,
  resetMocks: true,
  rootDir: path.resolve(__dirname),
  moduleNameMapper: {
    '^@oakjs/core(.+)$': '<rootDir>/../core/lib$1',
    '^@tests-utils$': path.resolve(__dirname, 'tests/utils.js'),
  },
  coveragePathIgnorePatterns: [
    'node_modules',
    'dist',
    'tests/',
    '^.+\\.styl$',
  ],
  transform: {
    '^.+\\.js$': 'babel-jest',
    '^.+\\.styl$': 'jest-css-modules-transform',
  },
  transformIgnorePatterns: [
    '/node_modules/(?!(uuid))',
  ],
  setupFilesAfterEnv: [
    './tests/env.js',
  ],
  snapshotResolver: path.resolve('.ci/config/snapshot-resolver.js'),
};
