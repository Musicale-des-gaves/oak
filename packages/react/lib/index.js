export {
  TextField,
  SelectField,
  DateField,
  ColorField,
  Dropdown,
  DropdownMenu,
  DropdownToggle,
  Toggle,
  Label,
  classNames,
} from '@junipero/react';

export { default as Builder } from './Builder';
export { default as Container } from './Container';
export { default as Catalogue } from './Catalogue';
export { default as Icon } from './Icon';
export { default as Text } from './Text';

export {
  useBuilder,
  useLogger,
} from './hooks';

export * from './addons';
