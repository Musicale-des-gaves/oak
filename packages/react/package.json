{
  "name": "@oakjs/react",
  "version": "3.0.0-alpha.28",
  "description": "ðŸŒ³ Modern, lightweight & modulable page builder",
  "main": "dist/oak-react.cjs.js",
  "jsnext:main": "dist/esm/index.js",
  "module": "dist/esm/index.js",
  "esnext": "src/index.js",
  "unpkg": "dist/oak-react.min.js",
  "cdn": "dist/oak-react.min.js",
  "types": "dist/oak-react.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/p3ol/oak.git",
    "directory": "packages/oak"
  },
  "author": "Ugo Stephant <ugo@poool.fr>",
  "contributors": [
    "Maxime Da Silva <maximed@poool.fr>"
  ],
  "license": "MIT",
  "sideEffects": false,
  "engines": {
    "node": ">= 14",
    "yarn": ">= 1"
  },
  "peerDependencies": {
    "react": "^18.0.0",
    "react-dom": "^18.0.0"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "7.20.13",
    "@floating-ui/react": "0.19.1",
    "@junipero/react": "3.0.0-beta.10",
    "@junipero/transitions": "3.0.0-beta.0",
    "core-js": "3.27.2",
    "preact": "10.11.3",
    "uuid": "9.0.0"
  },
  "scripts": {
    "clean": "rm -rf ./dist || true",
    "build": "yarn clean && rollup -c",
    "test": "jest"
  },
  "publishConfig": {
    "access": "public"
  },
  "exports": {
    ".": "./dist/esm/index.js",
    "./addons": "./dist/esm/addons.js",
    "./options": "./dist/esm/options.js"
  }
}
