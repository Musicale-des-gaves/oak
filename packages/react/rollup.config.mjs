import path from 'path';

import babel from '@rollup/plugin-babel';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import terser from '@rollup/plugin-terser';
import dts from 'rollup-plugin-dts';
import alias from '@rollup/plugin-alias';

const input = './lib/index.js';
const defaultOutput = './dist';
const name = 'oak-react';
const formats = ['umd', 'cjs', 'esm'];

const defaultExternals = [
  'react', 'react-dom',
];
const defaultGlobals = {
  react: 'React',
  'react-dom': 'ReactDOM',
};

const defaultPlugins = [
  babel({
    exclude: /node_modules\/(?!@oakjs)/,
    babelHelpers: 'runtime',
  }),
  alias({
    entries: [
      { find: '@oakjs/core', replacement: path.resolve('../core/lib') },
    ],
  }),
  resolve({
    rootDir: path.resolve('../../'),
  }),
  commonjs(),
  terser(),
];

const getConfig = (format, {
  output = defaultOutput,
  globals = defaultGlobals,
  external = defaultExternals,
} = {}) => ({
  input,
  plugins: [
    ...defaultPlugins,
  ],
  external,
  output: {
    ...(format === 'esm' ? {
      dir: `${output}/esm`,
      chunkFileNames: '[name].js',
    } : {
      file: `${output}/${name}.${format}.js`,
    }),
    format,
    name,
    sourcemap: true,
    globals,
    ...(format === 'esm' ? {
      manualChunks: id => {
        if (/packages\/core\/lib\/(\w+)\/index.js/.test(id)) {
          return path.parse(id).dir.split('/').pop();
        } else {
          return id.includes('node_modules') ? 'vendor' : path.parse(id).name;
        }
      },
    } : {}),
  },
});

export default [
  ...formats.map(f => getConfig(f)),
  getConfig('esm', {
    output: './dist/no-junipero',
    defaultExternals: [
      ...defaultExternals,
      '@junipero/react',
    ],
    defaultGlobals: {
      ...defaultGlobals,
      '@junipero/react': 'JuniperoReact',
    },
  }),
  {
    input: './lib/index.d.ts',
    output: [{ file: `dist/${name}.d.ts`, format: 'es' }],
    plugins: [dts()],
  },
];
