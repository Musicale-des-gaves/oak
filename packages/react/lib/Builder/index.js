import { forwardRef, useCallback, useRef, useImperativeHandle } from 'react';
import { classNames } from '@junipero/react';

import { BuilderContext } from '../contexts';
import { useRootBuilder } from '../hooks';
import Element from '../Element';
import Catalogue from '../Catalogue';

const Builder = forwardRef(({
  className,
  defaultValue,
  value,
  addons,
  rootBoundary,
  onChange,
  onImageUpload,
  ...opts
}, ref) => {
  const innerRef = useRef();
  const catalogueRef = useRef();
  const floatingsRef = useRef();
  const { builder, content } = useRootBuilder({
    content: value,
    defaultContent: defaultValue,
    addons,
    onChange,
    ...opts,
  });

  useImperativeHandle(ref, () => ({
    builder,
    content,
    isOak: true,
    catalogueRef,
    innerRef,
  }));

  const getContext = useCallback(() => ({
    builder,
    content,
    rootBoundary: rootBoundary?.current
      ? rootBoundary : { current: rootBoundary },
    onImageUpload,
    rootRef: innerRef,
    floatingsRef,
  }), [builder, content, rootBoundary, onImageUpload]);

  const onAppend = component => {
    catalogueRef.current?.close();
    builder.addElement({}, { component });
  };

  const onPrepend = component => {
    catalogueRef.current?.close();
    builder.addElement({}, { component, position: 'before' });
  };

  const onPaste = element => {
    catalogueRef.current?.close();
    builder.addElement(element, { resetIds: true });
  };

  return (
    <BuilderContext.Provider value={getContext()}>
      <div
        ref={innerRef}
        className={classNames(
          'oak builder',
          className
        )}
      >
        <div className="elements oak-flex oak-flex-col oak-gap-4">
          { content?.length > 0 && (
            <div className="add-element oak-flex oak-justify-center">
              <Catalogue
                ref={catalogueRef}
                onAppend={onPrepend}
                onPaste={onPaste}
              />
            </div>
          ) }

          { content?.map((element, i) => (
            <Element
              key={element.id || i}
              index={i}
              element={element}
            />
          )) }

          <div className="add-element oak-flex oak-justify-center">
            <Catalogue
              ref={catalogueRef}
              onAppend={onAppend}
              onPaste={onPaste}
            />
          </div>
        </div>

        <div className="floatings" ref={floatingsRef} />
      </div>
    </BuilderContext.Provider>
  );
});

Builder.displayName = 'Builder';

export default Builder;
