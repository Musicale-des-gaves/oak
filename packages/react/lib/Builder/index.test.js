import { render } from '@testing-library/react';

import { baseAddon } from '../addons';
import Builder from './index';

describe('<Builder />', () => {
  const getOptions = (...props) => {
    let i = 0;

    return {
      generateId: () => i++,
      ...props,
    };
  };

  it('should render', () => {
    const content = [
      { type: 'title', content: 'This is a title' },
      { type: 'text', content: 'This is a text' },
    ];

    const { container, unmount } = render(
      <Builder
        value={content}
        addons={[baseAddon()]}
        options={getOptions()}
      />
    );

    expect(container).toMatchSnapshot('Base content');
    unmount();
  });
});
