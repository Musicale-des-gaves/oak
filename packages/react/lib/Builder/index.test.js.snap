// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Builder /> should render 1`] = `
<div>
  <div
    class="oak builder"
  >
    <div
      class="elements oak-flex oak-flex-col oak-gap-4"
    >
      <div
        class="add-element oak-flex oak-justify-center"
      >
        <div
          class="catalogue"
        >
          <a
            class="handle oak-inline-flex oak-items-center oak-justify-center"
            draggable="false"
          >
            <i
              class="icon junipero-icons"
            >
              add
            </i>
          </a>
        </div>
      </div>
      <div
        class="oak element oak-flex-none type-title draggable"
        data-element-id="0"
        draggable="true"
      >
        <div
          class="inner oak-flex oak-gap-2 oak-p-4 oak-items-stretch"
        >
          <i
            class="icon junipero-icons"
          >
            title
          </i>
          <div
            class="element-info oak-flex-auto oak-flex oak-flex-col oak-gap-2 oak-justify-between"
          >
            <h6
              class="junipero oak-m-0"
            >
              Title
            </h6>
            <div
              class="element-content oak-flex-auto"
            >
              <h1
                class="junipero oak-m-0 sanitize-html !oak-text-4xl"
              >
                This is a title
              </h1>
            </div>
            <div
              class="props junipero extra !oak-text-alternate-text-color"
            >
              <span
                class="property"
              >
                <span
                  class="key"
                >
                  Type
                  : 
                </span>
                <span
                  class="value"
                >
                  Title 1 (h1)
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="options oak-flex oak-items-center"
        >
          <a
            class="option oak-flex oak-items-center oak-justify-center remove"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              close
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center duplicate"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              copy
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center copy"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              copy_file
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center oak-edit"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              pen
            </i>
          </a>
        </div>
      </div>
      <div
        class="oak element oak-flex-none type-text draggable"
        data-element-id="1"
        draggable="true"
      >
        <div
          class="inner oak-flex oak-gap-2 oak-p-4 oak-items-stretch"
        >
          <i
            class="icon junipero-icons"
          >
            multiline
          </i>
          <div
            class="element-info oak-flex-auto oak-flex oak-flex-col oak-gap-2 oak-justify-between"
          >
            <h6
              class="junipero oak-m-0"
            >
              Text
            </h6>
            <div
              class="element-content oak-flex-auto"
            >
              <div
                class="junipero sanitize-html"
              >
                This is a text
              </div>
            </div>
          </div>
        </div>
        <div
          class="options oak-flex oak-items-center"
        >
          <a
            class="option oak-flex oak-items-center oak-justify-center remove"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              close
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center duplicate"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              copy
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center copy"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              copy_file
            </i>
          </a>
          <a
            class="option oak-flex oak-items-center oak-justify-center oak-edit"
            draggable="false"
            href="#"
          >
            <i
              class="icon junipero-icons !oak-text-lg"
            >
              pen
            </i>
          </a>
        </div>
      </div>
      <div
        class="add-element oak-flex oak-justify-center"
      >
        <div
          class="catalogue"
        >
          <a
            class="handle oak-inline-flex oak-items-center oak-justify-center"
            draggable="false"
          >
            <i
              class="icon junipero-icons"
            >
              add
            </i>
          </a>
        </div>
      </div>
    </div>
    <div
      class="floatings"
    />
  </div>
</div>
`;
